ethereum/referencetests/build.gradle

```diff
@@ -13,16 +13,93 @@
+    )
+}
+
+def legacyBlockchainReferenceTests = tasks.register("legacyBlockchainReferenceTests")  {
+  inputs.files fileTree('reference-test/resources/LegacyTests/Constantinople/BlockchainTests'),
+    fileTree("$buildDir/generated/sources/reference-test/$name/java")
+  outputs.files "$buildDir/generated/sources/reference-test/$name/java"
+  generateTestFiles(
+    fileTree('src/reference-test/external-resources/LegacyTests/Constantinople/BlockchainTests'),
+    file("src/reference-test/templates/BlockchainReferenceTest.java.template"),
+    "LegacyTests",
+    "$buildDir/generated/sources/reference-test/$name/java/org/hyperledger/besu/ethereum/vm/blockchain",
+    "LegacyBlockchainReferenceTest",
+    ("BlockchainTests/InvalidBlocks/bcExpectSection") // exclude test for test filling tool
+    )
+}
+
+def generalstateReferenceTests = tasks.register("generalstateReferenceTests")  {
+  inputs.files fileTree("reference-test/resources/GeneralStateTests"),
+    fileTree("$buildDir/generated/sources/reference-test/$name/java")
+  outputs.files "$buildDir/generated/sources/reference-test/$name/java"
+  generateTestFiles(
+    fileTree("src/reference-test/external-resources/GeneralStateTests"),
+    file("src/reference-test/templates/GeneralStateReferenceTest.java.template"),
+    "GeneralStateTests",
+    "$buildDir/generated/sources/reference-test/$name/java/org/hyperledger/besu/ethereum/vm/generalstate",
+    "GeneralStateReferenceTest"
+    )
+}
+
+def legacyGeneralstateReferenceTests = tasks.register("legacyGeneralstateReferenceTests")  {
+  inputs.files fileTree('reference-test/resources/LegacyTests/Constantinople/GeneralStateTests'),
+    fileTree("$buildDir/generated/sources/reference-test/$name/java")
+  outputs.files "$buildDir/generated/sources/reference-test/$name/java"
+  generateTestFiles(
+    fileTree('src/reference-test/external-resources/LegacyTests/Constantinople/GeneralStateTests'),
+    file("src/reference-test/templates/GeneralStateReferenceTest.java.template"),
+    "LegacyTests",
+    "$buildDir/generated/sources/reference-test/$name/java/org/hyperledger/besu/ethereum/vm/generalstate",
+    "LegacyGeneralStateReferenceTest"
+    )
+}
+
+def generalstateRegressionReferenceTests = tasks.register("generalstateRegressionReferenceTests") {
+  inputs.files fileTree("reference-test/resources/regressions/generalstate"),
+    fileTree("$buildDir/generated/sources/reference-test/$name/java")
+  outputs.files "$buildDir/generated/sources/reference-test/$name/java"
+  generateTestFiles(
+    fileTree("src/reference-test/resources/regressions/generalstate"),
+    file("src/reference-test/templates/GeneralStateReferenceTest.java.template"),
+    "regressions",
+    "$buildDir/generated/sources/reference-test/$name/java/org/hyperledger/besu/ethereum/vm/generalstate",
+    "GeneralStateRegressionReferenceTest"
+    )
+}
+
 sourceSets {
-  test {
+  referenceTest {
```

Comment:
What advantage does a new reference test source set provide?  It will bring us further from "default" maven and gradle project setups, so I feel there needs to be a clear advantage to taking on this technical debt.

Reply:
I don't actually see this as technical debt here. It's an usual practice within Gradle to set different sourcesets for tests of different nature (e.g. _integrationTests_). But this also relates to what I replied to your other comment, it's to isolate the classpath from the _test_ task execution

Reply:
The reason I view it as technical debt is because source sets are a fairly poorly understood and fluid aspect of Gradle's object definitions (speaking from a 10 year perspective).  They (the gradle company) have a propensity for arbitrarily changing interactions with the source sets across newer versions of gradle, so my goal is to reduce the burden encountered when we need to support new versions of Gradle, and since upgrading Gradle is often required for new JDK versions we cannot easily say "stick with an older version."  This is movement from a "standard" layout to a non-standard layout.  We need to gain something from that for it to be a good idea.

Why do the classpaths even need to be isolated from the task execution?

Reply:
I'm with you about minimizing complexity within Gradle. The introduction of a different _sourceSet_ is intended to do exactly that. On `main`, the classpath of the **test** and the classpath of the **refereceTest** were hand picked by including/excluding specific `*.class` files, which looked a bit forced to me. Also, what I'm trying to do here with the _sourceSet_ is to remain as minimal as possible. This feature was introduced in version [0.8](https://docs.gradle.org/0.8/userguide/java_plugin.html#N112C9) and hasn't changed since then, so the scenario where they break compatibility seems pretty unlikely to me. I think that the introduction of this new _sourceSet_ would turn visible a preexisting de facto separation of classpaths, clear up the dependencies of the **referenceTests** and integrate seamlessly with IntelliJ. For me, the main point of this PR is that this allowed me to just update _these_ tests, instead of mixing those changes between the update of the `:ethereum:core` _unit_ tests

Reply:
If you are willing to take strong ownership over this I'll allow it.

Reply:
If we are going to commit to source sets for reference tests then the tasks that generate the tests (blockchainReferenceTests, generalstateReferenceTests, generalstateRegressionReferenceTests) should use source sets in their task definition instead of manually coded paths.

Reply:
Yes, I can take ownership over this and will take special care when we are willing to upgrade Gradle/JVM versions.
Regarding using _sourceSets_ for the other tasks, I don't see if they would suit fine given the nature of the tasks we are talking here. _sourceSets_ [are meant to be used](https://docs.gradle.org/7.4.2/userguide/building_java_projects.html#sec:java_source_sets) within a compile/build cycle of JVM-based languages, and the tasks here are for generating sources to be used **in** that cycle but for **referenceTest**. Even if we could find a way, I think we are getting enough integration between the new _sourceSet_ and the tasks by linking the `srcDirs` of it with the tasks _outputs_